### DISCUSSION TITLE
Syntax for suggesting alternative ingredients

### DISCUSSION THREAD

**Discussion by tobei** [cooklang/spec#68] [Ideas]
Some ingredients can be hard to find or unsuitable for some diet (maybe two separate usecases). Having a syntax to mark an ingredient as optional or suggesting alternative ingredient that can be substitued because they have similar taste, or vegan alternative, or not so important in the recipe could be useful.

That way the original recipe could feature all the original ingredients (e.g. Tabasco), but could feature a more general alternative as Spicy Sauce if Tabasco can't be found, or for baking 10g fresh yeast could have a 5g dry yeast alternative if fresh yeast isn't available.

**Comment by dubadub** [comment]
Thanks for idea @tobei. This will be definitely good addition to the language.

**Comment by jakeanq** [comment]
I think this would be very useful - I really like recipes that can be cooked with a range of different ingredients, for example a fruit pudding recipe that can be made with blackberries, blueberries or apples.

A syntax I think could work for this is separating the ingredients with a pipe symbol in a similar way to bitwise-OR operators in many languages etc and also very similar to what has been proposed here https://github.com/cooklang/spec/discussions/50#discussioncomment-1974093 :
```
Add @fresh yeast{10g}|@dry yeast{5g}.
```

Optionally a quick shorthand could also be added - useful when the quantities etc are the same for each option:
```
Add @tabasco sauce|spicy sauce{} to taste.
```

If an empty ingredient was used, the ingredient could be marked as optional, which covers #50:
```
If you like it spicy, add @red chilli{3}|@{}. -- Maybe a bit clunky?
-- and/or
If you like it spicy, add @red chilli|{3}.
```

For completeness I would suggest the same syntax be added to utensils:
```
Mix with a #spoon|eggbeater|electric drill{}
```

In terms of shopping lists, I would suggest that the first ingredient that is found in the shopping list config file be used - this supports the use cases of dietary requirements (specify only the things you can eat in the config file) and store availability.

Stretching further, I think it could become a part of a variants system, where one recipe can be cooked in several ways.  I will add a separate discussion for that though.

If you'd like a PR for this, let me know and I'll see what I can do :)

**Reply by digital-carver** [reply]
One disadvantage of that syntax is that it constrains the ingredients to have to be right next to each other in the text. 

Often, as the original post mentions, you want there to be an original suggested ingredient and alternates for specific cases eg. vegan diet, availability, etc. With the pipe syntax, you'd have to write them all together and then explain the why and when of the alternatives after the fact: 
```
Add @Tabasco sauce|spicy sauce{} to taste (Tabasco sauce is preferred, use 
Spicy sauce in case Tabasco isn't available)  
```

 which overall reads a bit awkward. And as a general rule, requiring the user to change the grammar/style of their writing for the sake of syntax seems like something to be avoided. 

A labeling system seems like a better way to go about it, allowing the writer to be free and flexible with their writing while allowing ingredients to be marked as alternatives for the same thing. Eg. 
```
Add @|sauce| Tabasco sauce{} to taste (in case Tabasco isn't available, 
use @|sauce| Spicy sauce{} instead". 
```

```
Pour @|milk| milk{1/2 % cup} into the mixture (as a vegan alternative, 
@|milk| almond milk{} can be used).
```

Writing those down gives me another idea: 
```
Add @Tabasco sauce{} to taste (in case Tabasco isn't available, 
use @/Tobasco sauce/ Spicy sauce{} instead". 
```
```
Pour @milk{1/2 % cup} into the mixture (as a vegan alternative, 
@/milk/ almond milk{1/2 % cup} can be used).
```

i.e. the main ingredient (the first alternative) is written as usual, and the later alternatives use the name of the main ingredient itself as the label. I've used slashes here as the delimiter for the label because that makes it resemble the text substitution syntax in Vim/perl/etc. (`s/old text/new text/`), which seems appropriate since the syntax denotes that the original ingredient is *substituted* by the alternative.

**Comment by FranGarc** [comment]
I'm thinking on a different approach. If you put the alternatives within some kind of marker with an "or" separator (say "[ ]") and adding the condition ("if you can't find", "if you prefer", etc) to the end, in a similar way to the [shorthand preparations notation](https://cooklang.org/docs/spec/#short-hand-preparations). 

It would need rewording work on the part of the user writing the recipe, as it's not a "natural" way of writing, but I'd say it'd pay the effort.

`Add [@Mild Ingredient{}|@ Spicy Alternative{}](if you like some fire in your food) `

Render: 

> Add Mild Ingredient or, if you like some fire in your food, Spicy Alternative.

`Boil in [@water{1L}|@veggie stock{1L}](if you want to add extra flavour) `

Render: 

> Boil in 1L water or, if you want to add extra flavour, 1L veggie stock.


Maybe it would also allow for multiple options

`Add [@Ingredient 1{}|@Alternative 1{}| @Alternative 2{}](if you can't find it|as a second alternative) `

Render:
> Add Ingredient 1 or, if you can't find it, Alternative 1 or, as a second alternative, Alternative 2.


I understand that different languages have different grammar and, therefore , can have different structures for this kind of conditionals/disjunctive clauses, but I can say that this works both in English and Spanish.  


I'd also say that the list of ingredients should reflect the fact that they're alternatives

> Ingredient 1
> Ingredient 2 (or Alternative 1 or Alternative 2)
> Ingredient 3

**Comment by tmlmt** [comment]
I feel like to give the most flexibility to cooks, you'd need two different additional syntax forms. 

1. @digital-carver's solution described [above](https://github.com/cooklang/spec/discussions/68#discussioncomment-2580517)

Syntax: `Pour @|milk|milk{1/2%cup}(lukewarm) into the mixture (as a vegan alternative, @|milk|almond milk{1/2%cup}(room temperature) can be used).`

Advantage:
- full control over text flow
- use case can be defined without having to invent new syntax

2. An in-place solution like @FranGarc's. 

Syntax: `Add @eggs{1}(white only)|@apple compot{2%tbsp}(smooth)[for a vegan version]` 

Advantage: possibility to choose how to handle the in-place display. Could be by concatenating the options with " or ", or by having a dropdown menu somewhere to choose which version you'd want to use

---

Several additional thoughts which would require even more syntax: 
- What if one ingredient needs to be substituted by several other ones?
- For option #2, should we give the possibility to define "versions" like (normal, vegan) for instance

**Comment by keithamus** [comment]
As I mentioned in https://github.com/cooklang/spec/discussions/75#discussioncomment-15706742 some ingredients might be a substitute for multiple in a recipe. As an example a cocktail recipe might call for "simple syrup, lemon juice" but this can be relaced with sour-mix, if it's on hand. A recipe for ramen might have a whole set of steps for building the soup base, but a pre-made soup base might be on hand.

Often times recipes are organised into ingredient categories, e.g. "Ingredients for the soup base". Though maybe interlinking recipes and having more component-based recipes is better?

**Reply by tmlmt** [reply]
> As I mentioned in [#75 (comment)](https://github.com/cooklang/spec/discussions/75#discussioncomment-15706742) some ingredients might be a substitute for multiple in a recipe. As an example a cocktail recipe might call for "simple syrup, lemon juice" but this can be relaced with sour-mix, if it's on hand. A recipe for ramen might have a whole set of steps for building the soup base, but a pre-made soup base might be on hand.
> 
> Often times recipes are organised into ingredient categories, e.g. "Ingredients for the soup base". Though maybe interlinking recipes and having more component-based recipes is better?

In that case you could also think about a case where 1 ingredient can be substituted by 2 others, or 2 ingredients by 3 others. What syntax would you suggest so that all cases can be covered?

The additional challenge with non-inline-alternatives (i.e. alternatives mentioned at a different place from where the original ingredient(s) is/are mentioned), is that you are then face with multiple rendering issues: what should be displayed by default? What should be displayed when an alternative is selected? etc. 

I've been been trying things out in a [playground](https://cooklang-parser.tmlmt.com/playground/) for the next major version of my own cooklang parser where I've introduced what I've described in [my previous comment](https://github.com/cooklang/spec/discussions/68#discussioncomment-15180132). Comments are welcome.

**Reply by keithamus** [reply]
Ahh yes good point. A recipe might call for "salted butter" which can be easily replicated with salt and butter, or "self raising flour" which can be replaced with flour and 2% baking powder. Perhaps anything more complex deserves to be it's own sub-recipe (for example hummus). I'm reminded of the Carl Sagan quote:

> If you wish to make an apple pie from scratch, you must first invent the universe.

**Reply by keithamus** [reply]
I think for my needs I'd prefer something like notes, but perhaps more formalised. In your example you call out vegan substitutes but if I wanted to make a vegan version of a recipe I'm likely to swap out many ingredients, and so we almost need a "recipe alternate". So maybe something in the front matter:


```md
title: Pancakes
variants:
	vegan:
		- @eggs{1}|@aquafaba{50ml}
		- @butter{50g}|@palm oil{50g}
		- @milk{200ml}|@oat milk{200ml}
	low_fat:
		- @butter{50g}|@low fat margarine{50g}
		- @milk{200ml}|@skimmed milk{200ml}
---

Combine @eggs{1}, @butter{50g}, @milk{200ml} in a jar and blend.

Pour 50g into a pan.
```

Or maybe something which can replace a step, allowing you to give more detail if needed.

```md
title: Pancakes
---

Combine @eggs{1}, @butter{50g}, @milk{200ml} in a jar and blend.

> ![vegan]
> Combine @aquafaba{50ml}, @palm oil{50g}, @oat milk{200ml} in a jar and blend.

> ![low fat]
> Combine @eggs{1}(white only), @low fat margarine{50g}, @skimmed milk{200ml} in a jar and blend.

Pour 50g into a pan.
```

**Reply by tmlmt** [reply]
What's missing is a way to determine what is a substitute for what. 

In your second example, how would we know which of `Combine @eggs{1} ...` or `Pour 50g into a pan` the vegan step replaces?

**Reply by romainhild** [reply]
you could have different lists of ingredients for different versions of the recipe (original, vegan, low fat, ...). Only the ingredients that appear on the steps of the version appear on the list.

**Reply by tmlmt** [reply]
Yes. 

I've answered the questions I was asking in my two comments above and implemented things in v3 of my own parser, in two complementary/alternative ways:
- grouped alternatives with group and subgroup keys, see [guide](https://cooklang-parser.tmlmt.com/v3/guide-extensions.html#grouped-alternatives)
- recipe variants, see https://github.com/cooklang/spec/discussions/75#discussioncomment-16377026

You can test this out in this [playground](https://cooklang-parser.tmlmt.com/playground/)

Could serve as proof of concept / inspiration for the official cooklang spec and toolchain.
